database/spanner.go

```diff
@@ -24,3 +29,17 @@ func SpannerMigrationDriver(cfg SpannerConfig, databaseName string) (database.Dr
 	s := migspanner.Spanner{}
 	return s.Open(fmt.Sprintf("spanner://projects/%s/instances/%s/databases/%s?x-migrations-table=spanner_schema_migrations&x-clean-statements=true", cfg.Project, cfg.Instance, databaseName))
 }
+
+// SpannerUniqueViolation returns true when the provided error matches the Spanner code
+// for duplicate entries (violating a unique table constraint).
+// Refer to https://cloud.google.com/spanner/docs/error-codes for Spanner error definitions,
+// and https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto for error codes
+func SpannerUniqueViolation(err error) bool {
+	match := strings.Contains(err.Error(), "Failed to insert row with primary key")
+
+	var apiErr *apierror.APIError
+	if errors.As(err, &apiErr) {
+		return match || spanner.ErrCode(apiErr) == codes.AlreadyExists
```

Comment:
I think you can remove `match` in this case if you're already doing `spanner.ErrCode`

Reply:
In the [example](https://gist.github.com/gyao852/0319adb1c5bb521194147a8ff23b3769) they recommended to cast to `APIError` to get full information (like `Desc` and `additionalInformation` I think) but yeah I guess if we only need the `ErrCode` part I can just unwrap directly.

Reply:
done